src/wizards/newspack/views/settings/additional-brands/index.tsx

```diff
@@ -0,0 +1,235 @@
+/**
+ * Additional Brands page.
+ */
+
+/**
+ * WordPress dependencies.
+ */
+import { __ } from '@wordpress/i18n';
+import { addQueryArgs } from '@wordpress/url';
+import { useState, useEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies.
+ */
+import Brands from './brands';
+import BrandUpsert from './brand-upsert';
+import WizardsTab from '../../../../wizards-tab';
+import WizardSection from '../../../../wizards-section';
+import { Router, utils } from '../../../../../components/src';
+import { useWizardApiFetch } from '../../../../hooks/use-wizard-api-fetch';
+
+const { Route, Switch, useHistory, useRouteMatch } = Router;
+
+export default function AdditionalBrands() {
+	const { wizardApiFetch, isFetching, cache } = useWizardApiFetch(
+		'newspack-settings/additional-brands'
+	);
+
+	const brandsCache = cache( '/wp/v2/brand' );
+
+	const [ brands, setBrands ] = useState< Brand[] >( [] );
+	const history = useHistory();
+	const { path } = useRouteMatch();
+
+	/**
+	 * Cache brands data.
+	 */
+	useEffect( () => {
+		brandsCache.set( brands );
+	}, [ brands ] );
+
+	const wizardScreenProps = {
+		isFetching,
+		headerText: __( 'Brands', 'newspack-plugin' ),
+		subHeaderText: __( 'Configure brands settings', 'newspack-plugin' ),
+	};
+
+	/**
+	 * Fetching brands data.
+	 */
+	const fetchBrands = () => {
+		wizardApiFetch< Brand[] >(
+			{
+				path: addQueryArgs( '/wp/v2/brand', { per_page: 100 } ),
+			},
+			{
+				onSuccess( response ) {
+					setBrands(
+						response.map( ( brand: Brand ) => ( {
+							...brand,
+							meta: {
+								...brand.meta,
+								_theme_colors: brand.meta._theme_colors?.length
+									? brand.meta._theme_colors
+									: [],
+								_menus: brand.meta._menus?.length
+									? brand.meta._menus
+									: [],
+							},
+						} ) )
+					);
+				},
+			}
+		);
+	};
+
+	const upsertBrand = ( brandId: number, brand: Brand ) => {
+		// BrandId is NaN when inserting new brand.
+		wizardApiFetch< Brand >(
+			{
+				path: brandId ? `/wp/v2/brand/${ brandId }` : '/wp/v2/brand',
+				method: 'POST',
+				data: {
+					...brand,
+					meta: {
+						...brand.meta,
+						...( brand.meta._logo && {
+							_logo:
+								brand.meta._logo instanceof Object
+									? brand.meta._logo.id
+									: brand.meta._logo,
+						} ),
+					},
+				},
+			},
+			{
+				onSuccess( result ) {
```

Comment:
When updating, we should rehydrate the local store with `result` so it's guaranteed to reflect what's in the DB.

When creating a brand with an empty slug, the slug will be automatically generated and returned by the API. Currently, this will only be visible in the UI if I do a page refresh.

Reply:
Can you share some reproducible steps, I'm unable to replicate this, unless I've misunderstood?

Here is a screen-capture of me adding a brand and it appearing in the brand list after clicking "Save" (I also navigate to a different tab and back to confirm the data is persistent).

https://github.com/user-attachments/assets/39e723e7-30e0-40f8-a04c-2ff8f5840e91

Here is another screen capture showing a persistent update:

https://github.com/user-attachments/assets/3c35f889-9388-443a-9134-4914ba072607

The code that takes care of this is:
https://github.com/Automattic/newspack-plugin/blob/75833a3372b61af7bf1d0ae927aabc0acc145db0/src/wizards/newspack/views/settings/additional-brands/index.tsx#L39-L41

Reply:
I noticed that when editing the brand, the slug is auto-generated via:

https://github.com/Automattic/newspack-plugin/blob/75833a3372b61af7bf1d0ae927aabc0acc145db0/src/wizards/newspack/views/settings/additional-brands/brand-upsert.tsx#L149-L153

I'm not sure why this did not trigger during my testing, but it is now.

Another example to test what I mean is creating 2 brands with different names but the same slug. The second brand will have its slug modified with the `-2` suffix on the backend, which will not be reflected on the front-end.

Even though this is a very edge case, it shows that the backend can make changes to the data and the front-end should be aware of that.

It seems like changing line 101 of `src/wizards/newspack/views/settings/additional-brands/index.tsx` to `const newBrand = { ...result, id: result.id };` should do the trick, but I could be missing something.

Reply:
> Even though this is a very edge case, it shows that the backend can make changes to the data and the front-end should be aware of that.

@miguelpeixe 100% agree! I went ahead and ran two test scenarios and fixed both issues I encountered:

1. Two brands exist. Attempt to save one brand using the other brands slug e.g. "slug-one". 
   - This resulted in an error from the backend - "The slug &#8220;slug-one&#8221; is already in use by another term."
   - Solution: Added error message handling to the upsert component - c10e4d73f5e0a0ce8cfeabf943a5d1ef395962c2.
![Screenshot 2024-10-23 at 09 42 48](https://github.com/user-attachments/assets/520c5e74-3332-4905-947d-6f86a77042a3)

2. One brand exists. A new brand is created using the existing brands slug e.g. "slug-one" (these scenario closely resemble the experiences you've mentioned).
   - This saved without error.
   - After save the brand appears to use the same "slug-one".
   - BUT: when page is refreshed the slug changes to "slug-one-2".
   - Solution: Use `result.slug` when updating cache - 46b001885e6e970db5ad4af4822cd48be05e62f5.
![Screenshot 2024-10-23 at 10 05 24](https://github.com/user-attachments/assets/3c1d6160-5bec-4521-bbab-4dcadf791d58)

Reply:
Nice catch with the error handling, works nicely!

> Solution: Use result.slug when updating cache

This works for this scenario, but is there a reason not to use the entire backend response object here? The UI has to match the full response, controlling it by property adds some maintainability burden for future developments.

Reply:
> but is there a reason not to use the entire backend response object here

No particular reason. I was trying to avoid a bigger footprint in changes from [original implementation](https://github.com/Automattic/newspack-multibranded-site/blob/1ddbba3899889320f1cf4f5b21cd939865906bb8/src/admin/views/brands/index.js#L62). I have updated to use entire `result` d95e89daf575711299af7cfe6455460ad7ecd03e

Reply:
Ah, I see! Thank you for clarifying and revising the approach.